Add formatOnSave and ability to pass syntax

RobbieTheWagner opened the pull request:
Resolves #2 and resolves #5

RobbieTheWagner committed:
Add ability to pass syntax

RobbieTheWagner committed:
Add formatOnSave and ability to pass syntax

Resolves #2 and resolves #5

sindresorhus commented on index.js:
```diff
@@ -83,9 +94,24 @@ export const config = {
 		],
 		default: 80
+	},
+	syntax: {
```
As commented in the original issue, it shouldn't be an option, but just handled automagically.

RobbieTheWagner replied:
Did not see that comment

sindresorhus commented on index.js:
```diff
@@ -3,6 +3,12 @@ import postcss from 'postcss';
 
+const SUPPORTED_SCOPES = [
+	'source.css',
+	'source.css.less',
```
Is LESS supported by Perfectionist?

RobbieTheWagner replied:
@sindresorhus I believe it will work with anything there is a Post-CSS plugin for, so I think postcss-less would work perhaps? I'm not sure if the `syntax` param only supports `scss` or if `less` would work too.

sindresorhus replied:
@rwwagner90 Then Perfectionist would need to add that parser first.

sindresorhus commented on index.js:
```diff
@@ -56,6 +62,11 @@ export const config = {
 		]
 	},
+	formatOnSave: {
+		description: 'Check the box to automatically format on save',
```
This is just repeating the option name in a more verbose form. Not needed.

RobbieTheWagner replied:
I see nothing wrong with the description, but if you insist.

RobbieTheWagner committed:
Make sass automagically apply syntax, remove formatOnSave description

RobbieTheWagner committed:
Remove less option

RobbieTheWagner committed:
Make config a let

RobbieTheWagner commented:
@sindresorhus okay, I made scss apply the syntax automagically, I removed less as a supported scope, and I removed the verbose description you requested. Any other issues?

RobbieTheWagner committed:
Fix lint issues

sindresorhus commented on index.js:
```diff
@@ -3,6 +3,11 @@ import postcss from 'postcss';
 import perfectionist from 'perfectionist';
 import postcssSafeParser from 'postcss-safe-parser';
 
+const SUPPORTED_SCOPES = [
```
Nitpick, but this would be better as a [Set](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Set).

RobbieTheWagner replied:
I copied this from your code https://github.com/sindresorhus/atom-esformatter/blob/b7eee38f1c51a0a4cbb96c0cb79862a72746def9/index.js#L5

It's a `const` there.

sindresorhus replied:
It would be better as a `Set` there too.

> It's a `const` there.

My request has nothing to do with the `const`.

RobbieTheWagner replied:
I've never used a `Set` myself. Can we just leave it as a `const` for now?

RobbieTheWagner replied:
I actually just went ahead and looked up how to use a `Set` and I, hopefully, made the change you were looking for.

sindresorhus commented on index.js:
```diff
@@ -12,8 +17,15 @@ function init() {
+	const config = atom.config.get('perfectionist');
 
-	postcss(perfectionist(atom.config.get('perfectionist'))).process(text, {
+	if (editor.getGrammar().scopeName === 'source.css.scss') {
```
Per original issue, it should be the inverse. SCSS unless there's a `.css` extension:

> This is how I did SCSS support in atom-autoprefixer: https://github.com/sindresorhus/atom-autoprefixer/blob/99bac7cf2f44a41dc552143441240e4c4a4cf9b5/index.js#L17-L22 **Pretty much just use the SCSS parser unless the file is .css. That makes it work for SCSS on unsaved buffers too.**

RobbieTheWagner replied:
Okay, I made this change, but I don't understand what you mean by working with unsaved buffers. If it is unsaved, I assume the type is not known is what you are saying? If that is true, we're now running the scss syntax when there is an unsaved css file, which would be incorrect, right?

sindresorhus replied:
@rwwagner90 SCSS is a superset of CSS, so all CSS works in the SCSS parser, but not the other way around, by using SCSS on unsaved buffers we make it work for both CSS and SCSS.

RobbieTheWagner committed:
Set syntax to scss by default and delete it if not scss

RobbieTheWagner committed:
Check against css

RobbieTheWagner commented:
@sindresorhus any more tweaks you'd like me to make?

sindresorhus commented on index.js:
```diff
@@ -12,8 +17,15 @@ function init() {
+		delete config.syntax;
+	}
+
+	postcss(perfectionist()).process(text, {
```
You're not no longer passing the config to `perfectionist`.

sindresorhus commented on index.js:
```diff
@@ -12,8 +17,15 @@ function init() {
-	postcss(perfectionist(atom.config.get('perfectionist'))).process(text, {
+	if (editor.getGrammar().scopeName === 'source.css') {
+		delete config.syntax;
+	}
```
Would be better like this:

``` js
if (editor.getGrammar().scopeName !== 'source.css') {
    config.syntax = 'scss';
}
```

RobbieTheWagner replied:
I was doing that before, essentially, and you said it should default to scss all the time to work with unsaved buffers. Not sure how this is any different than what I had previously.

sindresorhus replied:
You were not. This is just a different version of the above. In https://github.com/sindresorhus/atom-perfectionist/pull/7/commits/2d21b7fc2bd9944b5239d77153402f88a3baa41b#diff-168726dbe96b3ce427e7fedce31bb0bcR23, you were setting the `syntax` option when it's a SCSS file. This is setting the SCSS when it's **not** a CSS file. That's an important difference and what makes it work for SCSS on unsaved buffers.

RobbieTheWagner committed:
Pass in mistakenly omitted config

RobbieTheWagner committed:
Make requested change

RobbieTheWagner committed:
Fix lint

RobbieTheWagner commented:
@sindresorhus I made the changes you requested. Is there anything else?

RobbieTheWagner committed:
Change SUPPORTED_SCOPES to a Set

sindresorhus commented:
Looks good. Thank you :)

RobbieTheWagner commented:
You're welcome! Thanks for merging. I'm not sure how releasing versions of atom packages works, but if you could bump and release that would be awesome 😄

sindresorhus commented:
Already done 20 min ago ;)

https://github.com/sindresorhus/atom-perfectionist/commit/29aad8cc036a0c6196a37e42dbb614ec91d0e671

RobbieTheWagner commented:
Hmm, I just grabbed the latest version, and it doesn't seem to do anything on save 😞

sindresorhus commented:
Did you enable the `Format on Save` option?

RobbieTheWagner commented:
Yeah, I did. I suppose somewhere in all the changes things stopped working. I know it worked with the initial PR

RobbieTheWagner commented:
Just kidding, now it works magically. No idea what was happening.

RobbieTheWagner commented:
It does seem to delete all comments though. I suppose that means the `scss` syntax does not work?

sindresorhus commented:
Dunno. I only tested it without comments.

RobbieTheWagner commented:
@ben-eb is that the behavior we would see, if `syntax` was not correctly set to `scss`? The one line comments would be removed entirely? Trying to see if this is an issue with this package or perfectionist itself.

ben-eb commented:
Not necessarily, but you might experience flaky behaviour if you try to parse single-line comments with the default PostCSS parser. The reason is because single-line comments are not a CSS construct, hence the special parser. 😄

RobbieTheWagner commented:
@ben-eb yeah, I understand. I'm just trying to figure out which package the error lies in, with atom-perfectionist, or perfectionist itself.

sindresorhus commented:
@rwwagner90 Tracing the code would quickly answer that ;)

Quickly looked into it and the issue is that the `syntax` option only works when using `perfectionist.process` directly, not when using it as a `postcss` plugin like we do here. I just made this plugin handle the SCSS parser itself, which was fairly easy. Try 1.1.1.

RobbieTheWagner commented:
@sindresorhus seems to turn `//` comments into `/**/` comments. Is that intended? Much better than deleting them altogether though! 👍 

I am noticing something weird though. The format on save option seems to only work if you have manually run perfectionist from command palette at least once. If you never have, saving doesn't do anything. Should I open an issue for this?

sindresorhus commented:
> seems to turn // comments into /**/ comments. Is that intended? Much better than deleting them altogether though! 

I noticed that too. It seems to be a weakness in https://github.com/postcss/postcss-scss or Perfectionist. I've manually verified it passes the correct parser now.

> I am noticing something weird though. The format on save option seems to only work if you have manually run perfectionist from command palette at least once. If you never have, saving doesn't do anything. Should I open an issue for this?

Fixed in 1.1.2.

ben-eb commented:
This should only affect the `compact` and `compressed` modes. Perhaps the checking needs to be more robust? https://github.com/ben-eb/perfectionist/blob/master/src/index.js#L25-L27

sindresorhus commented:
I dunno how Perfectionist works, but I had `expanded` in the config.

ben-eb commented:
OK. If it's a bug then I'll need more details and a reproducible test case.

RobbieTheWagner commented:
@sindresorhus it does now work on all saves, even the first save, but now it is deleting single line comments again :(
